blog/2021-05-29-using-rust.md

```diff
@@ -0,0 +1,74 @@
+- The food tip, inference results and the image hash form the final response.
+
+## Stack decisions: Rust and haskell
+
+Haskell is used for the client facing API, and it's for other ad-hoc tasks such as reading/writing to a database among others.
+
+Rust is used for image preprocessing, model inference and sending the results back. Sending of results was done differently in the two approaches outlined below. Rust was chosen for it's high efficiency, small executable footprint, and absence of a garbage collector. It also had strong type system, speed & relatively actively maintained Tensorflow (client) library.
+
+# Introduction and pitfalls of the existing architecture
+
+![](/images/v2-arch-diagram.png)
+The internal organization of the rust service in this architecture is outlined above.
+There were 3 main parts, all running concurrently on 3 separate tokio runtimes - namely polling sqs, preprocessing and inferring from the images, and cleanup tasks.
+
+The external processes related to this architecture are outlined below.
+![](/images/v2-external-arch-diagram.png)
+
+The main gripe we had was in the `S3` to `SQS` upload event notification. In our experiments, it was [very slow](https://github.com/Holmusk/aws_benchmarks/tree/master/s3_event_to_sqs), and we aim for the service to have a very low latency, with the goal being that every image that comes into the system should be scored/rated in under 1 second. Because of the way the system was designed, this meant that we'd need a pretty big makeover on the rust side, and some tweaks on the haskell side. This is also mentioned in the [AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html), where they point that `Typically, event notifications are delivered in seconds but can sometimes take a minute or longer`.
+
+# The new architecture
+
+Like mentioned before, the main reason for redesigning the architecture was to avoid the `S3` to `SQS` upload event notification as speed was of high importance. In the process, we found out that we actually simplified it, by removing unnecessary moving parts.
+
+Internally, the rust service now also has a webserver. The client facing API proxies the requests it receives to the rust webservers via a load balancer. In this version, we completely eliminate the use of SQS.
+
+## Rust webserver internals
+
+![](/images/v3-arch-diagram.png)
+
+We use a [`warp`](https://github.com/seanmonstar/warp) for the webserver in rust.
+
+Because we have multiple tokio runtimes, the messages are passed between them using a bounded channel. We also have an extra batching step, which converts a bunch of independent requests into a batch. The batching step is important as the model inference is efficient for an image batch as compared to a single image. Also, with a batch, we can parallelize the image download and preprocessing steps for extra speed gains.
```

Comment:
```suggestion
We have 3 tokio runtimes running simultaneous and somewhat independently of one another. These tokio runtimes communicate with each other using messages that are passed between them using bounded channels. 
```

Let me know if the above sounds factually or technically incorrect. 

- Link to what a `channel` is 
- Link to footer where you can elaborate what "message"s are passed between channels i.e. you can say it is just a custom Rust type containing relevant information about an image